### ISSUE TITLE
new Check: ConstructorsDeclarationGrouping to check the grouping of overloaded constructors

### ISSUE THREAD

**Issue by Zopsss** [checkstyle/checkstyle#14726]
initially discuses at https://github.com/checkstyle/checkstyle/issues/14415 and https://github.com/checkstyle/checkstyle/pull/14398
Reason of new check creation is desire to cover Google style request for this - https://github.com/checkstyle/checkstyle/issues/14273

I have downloaded the latest cli from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

**How it works Now:**
Checkstyle have a module called  [OverloadMethodsDeclarationOrder](https://checkstyle.org/checks/coding/overloadmethodsdeclarationorder.html) which checks that if the overloaded methods are grouped together or not, if there is another method between overloaded methods then it'll give you an error. Currently, in checkstyle there is no module to check the grouping of overloaded constructors in a class, enum, abstract class, records. This new module will check that the overloaded constructors are grouped together or not. If there is another method or anything else between the constructors ( expect comments ), it will give an error.

**How I want check to work**
```xml
$ cat config.xml 
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
        "https://checkstyle.org/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="ConstructorsDeclarationGrouping"/>
    </module>
</module>
```

**Test.java**
```java
$ cat Test.java 
public class Test {
    public Test(){}

    public Test(int a){}

    public void testing() {}

    public Test(int a, int b) {}

}
```

**Expected output**
The module should check the grouping of overloaded constructors and give an error if they don't follow the rule. It should give an error something like this:

```
$ java -jar checkstyle-10.13.0-all.jar -c config.xml Test.java 
Starting audit...
[ERROR] C:\checkstyle testing\Test.java:8:3: All overloaded constructors should be grouped together. Previous overloaded constructor located at line ''4''.[ConstructorsDeclarationGrouping]
Audit done.
Checkstyle ends with 1 errors.
```

---
**Some other examples showing how the check should work**
Example 1: This is fine
```java
public class Example1 {
    Example1 () { }

    Example1 (int a) { }

    Example1 (int a, int b) { }

    void testing () { }
}
```

Example 2: This is fine as we do not required to have overloaded constructors in a specific order like in ascending order based on the number of parameters or something similar to that. We only care about that if the overloaded constructors are grouped together or not, we don't check their order based on number of parameters or anything.

```java
public class Example2 {
    Example2 (int a, String str) { }

    Example2 () { }

    Example2 (int a) { }
}
```

Example 3: This will give error, constructors are not grouped together
```java
enum Example3 {
  TEST1, TEST2, TEST3;

  Example3() { }

  Example3(String a) { }

  void testing() { } 

  Example3(int a) { } // violation, all overloaded constructors should be grouped together 
}
```

Example 4: This is fine
```java
record Example4(int x, int y) {
  Example4() {
    this(0, 0);
  }

  Example4(int x, int y) {
    this.x = x;
    this.y = y;
  }

  Example4(int x) {
    this(x, 0);
  }

  Example4(int x, int y, int z) {
    this(x, y + z);
  }
}
```

Example 5: This will give error, constructors are not grouped together
```java
record Example5(int x, int y) {

  Example5(int x) {
    this(x, 0);
  }

  Example5() {
    this(0, 0);
  }

  void testing() { } 

  Example5(int x, int y, int z) { // violation, all overloaded constructors should be grouped together
    this(x, y + z);
  }
}
```

Example 6: This is fine
```java
public abstract class Example6 {

    int a = 19;

    Example6 () {}

    Example6 (String a, String b) {}

    Example6 (int a) {}
}
```

Example 7: This will give error, constructors are not grouped together
```java
public abstract class Example7 {

    Example7 () {}

    Example7 (String a, String b) {}

    int a = 19;

    Example7 (int a) {} // violation, all overloaded constructors should be grouped together
}
```

---

### **Solution**
The module should be able to verify that all overloaded constructors are grouped together in a class, enum, records, etc. The overloaded constructors should be strictly grouped together, if there is a method, variables, etc between them then it will give an error.

**Comment by nrmancuso** [comment]
@Zopsss please update https://github.com/checkstyle/checkstyle/issues/14726#:~:text=Example%206%3A%20This%20will%20give%20error%2C%20constructors%20are%20not%20grouped%20together to be compilable to make sure we are on the same page.

**Comment by Zopsss** [comment]
> @Zopsss please update example 6 to be compilable to make sure we are on the same page.

done thanks for pointing this out.
---
@rnveach can you please review this issue?

**Comment by nrmancuso** [comment]
I am good to approve

**Comment by Zopsss** [comment]
@rnveach ping

Can you please review this issue? The other maintainers has approved this issue, We need your assurance for the approval.

**Comment by Zopsss** [comment]
Closes via #14844
